File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;
class Solution {
    public int solution(String begin, String target, String[] words) {
        if(!Arrays.asList(words).contains(target))
            return 0;
        Queue<String> q = new LinkedList<>();
        Queue<Integer> depth = new LinkedList<>();
        boolean[] visited = new boolean[words.length];
        q.offer(begin);
        depth.offer(0);
        
        while(!q.isEmpty()) {
            String current = q.poll();
            int d = depth.poll();
            if(current.equals(target))
                return d;
            for(int i = 0; i < words.length; i++) {
                if(!visited[i] && canchange(current, words[i])) {
                    visited[i] = true;
                    q.offer(words[i]);
                    depth.offer(d + 1);
                }
            }
        }
        return 0;
    }
    public boolean canchange(String a, String b) {
        int diff = 0;
        for(int i = 0; i < a.length(); i++) {
         if(a.charAt(i) != b.charAt(i))
             diff++;
        }
        return diff == 1;
    }
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <string>
#include <vector>
#include <queue>
#include <sstream>
#include <iomanip>

using namespace std;

bool canChange(const string& a, const string& b) {
    int diff = 0;
    
    for(int i = 0; i < a.size(); i++) {
        if(a[i] != b[i]) {
            diff++;
        
            if(diff > 1) {
                return false;
            }
        }
    }
    return diff == 1;
}

int solution(string begin, string target, vector<string> words) {
    int n = words.size();
    vector<bool> visited(n, false);
    
    queue<pair<string, int>> q;
    q.push({begin, 0});
    
    while(!q.empty()) {
        auto [cur, cnt] = q.front();
        q.pop();
        
        if(cur == target) return cnt;
        
        for(int i = 0; i < n; i++) {
            if(visited[i]) continue;
            
            if(canChange(cur, words[i])) {
                visited[i] = true;
                q.push({words[i], cnt + 1});
            }
        }
    }
    return 0;
}